client/my-sites/hosting/github/github-connect-card/index.tsx

```diff
@@ -1,16 +1,106 @@
-import { Card } from '@automattic/components';
-import { useI18n } from '@wordpress/react-i18n';
+import { Button, Card, Spinner } from '@automattic/components';
+import { __ } from '@wordpress/i18n';
+import { useEffect, useState } from 'react';
+import { useSelector } from 'react-redux';
 import CardHeading from 'calypso/components/card-heading';
-import iconGitHub from '../github.svg';
+import FormFieldset from 'calypso/components/forms/form-fieldset';
+import FormSelect from 'calypso/components/forms/form-select';
+import FormTextInput from 'calypso/components/forms/form-text-input';
+import SocialLogo from 'calypso/components/social-logo';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+import { RepoBranch, useGithubBranches } from './use-github-branches';
+import { Repo, useGithubRepos } from './use-github-repos';
+
+import './style.scss';
 
-// todo - just a placeholder for now as the implementation of this
-// component falls under another PR.
 export const GithubConnectCard = () => {
-	const { __ } = useI18n();
+	const siteId = useSelector( getSelectedSiteId );
+
+	const [ selectedRepo, setSelectedRepo ] = useState< Repo >();
+	const [ selectedBranch, setSelectedBranch ] = useState< RepoBranch >();
+	const { data: repos, isLoading: isLoadingRepos } = useGithubRepos( siteId );
+	const { data: branches, isLoading: isLoadingBranches } = useGithubBranches(
+		siteId,
+		selectedRepo?.name
+	);
+
+	useEffect( () => {
+		if ( repos ) {
+			setSelectedRepo( selectedRepo || repos[ 0 ] );
+		}
+		if ( branches ) {
+			setSelectedBranch( branches[ 0 ] );
+		}
+	}, [ repos, branches, selectedRepo ] );
```

Comment:
I think you need to do the `selectedBranch || branches[ 0 ]` trick here too?

We shouldn't need to have a `useEffect` here. Updating state based on some other state change causes a cascade of updates. The list of repos will load, causing a rerender, then this effect callback runs, which updates the state, which causes a rerender.

The idiomatic ways to remove the `useEffect` would be either to [calculate what the selected value should be](https://beta.reactjs.org/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state) i.e. when setting the `value` prop something like this:

```tsx
value={ selectedRepo?.full_name || repos?.[0].full_name }
```

Or we could do this trick where [the state is updated directly in the render path](https://beta.reactjs.org/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) (I don't see this pattern very often).

Or we could pass an `onSuccess` handler to the query hooks, and we set the selected repo on success. I actually like the pattern of exposing all the ReactQuery options as one of the arguments to `useGithubRepos`, and I also think it's much clearer to just say "on success set the default dropdown selection" than to have to deal with `useEffect`

Reply:
The initial value for these two properties, and the base path, should come from site options. I am using these constants:

```php
define( 'GITHUB_HOSTING_REPO_NAME', 'GITHUB_HOSTING_REPO_NAME' );
define( 'GITHUB_HOSTING_BRANCH_NAME', 'GITHUB_HOSTING_BRANCH_NAME' );
define( 'GITHUB_HOSTING_BASE_PATH', 'GITHUB_HOSTING_BASE_PATH' );
```
>We shouldn't need to have a useEffect here.

And then should be able to remove this `useEffect` as mentioned

Reply:
@mpkelly I don't think we need to use those for the connect branch card which is only shown after user is authorized, those will be shown on the disconnect branch card at which point I assume thos site options will also be cleared and the connect branch card will also fetch list of repo and branches again.

@p-jackson I knew something smelled with that `useEffect` since I was using ReactQuery which internally has `useEffect`.  I even had a discussion with chatGPT on it 😆 asking it if it's best practice to use useEffect while using react query. 

> we could pass an onSuccess handler to the query hooks
I like this approach better.